### PR TITLE
#17980: MDS client blacklisting and blacklist on eviction

### PR DESCRIPTION
**Pull request by jcsp**
This is the big rework/strengthening of eviction:
 * Evicting a client from one MDS daemon evicts it from all
 * Once a client is evicted it cannot reconnect to any MDS daemon
 * Evicted clients are blacklisted in the OSDMap, so they can no longer write data to OSDs
 * Clients notice their own eviction and do some teardown so that they can error out any operations in flight and be unmountable.

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -677,13 +684,25 @@ void Server::find_idle_sessions()
...
     mds->clog->info() << "closing stale session " << session->info.inst
 	<< " after " << age;
-    dout(10) << "autoclosing stale session " << session->info.inst << " last " << session->last_cap_renew << dendl;
-    kill_session(session, NULL);
+    dout(10) << "autoclosing stale session " << session->info.inst << " last "
+             << session->last_cap_renew << dendl;
+
+    if (g_conf->mds_session_blacklist_on_timeout) {
+      std::stringstream ss;
+      mds->kill_session(session->info.inst.name.num(), false, true,
```
Comment (jcsp) [inline src/mds/Server.cc]: I should rename this!

File: src/client/Client.cc [hunk 2]
```diff
@@ -1666,6 +1667,11 @@ int Client::make_request(MetaRequest *request,
     if (request->aborted())
       break;
 
+    if (blacklisted) {
+      request->abort(-EIO);
```
Comment (jcsp) [inline src/client/Client.cc]: Let's replace this with ESHUTDOWN

**Comment by batrick** [comment]
@jcsp, FYI this has conflicts now.

File: src/osdc/Objecter.cc [hunk 3]
```diff
@@ -1361,6 +1370,58 @@ void Objecter::handle_osd_map(MOSDMap *m)
...
+  std::set<entity_addr_t> old_set;
+  std::set<entity_addr_t> new_set;
+
+  old_osd_map.get_blacklist(&old_set);
+  new_osd_map.get_blacklist(&new_set);
+
+  std::set<entity_addr_t> delta_set;
+  std::set_difference(
+      new_set.begin(), old_set.begin(), new_set.end(), old_set.end(),
+      std::inserter(delta_set, delta_set.begin()));
```
Comment (ukernel) [inline src/osdc/Objecter.cc]: see http://en.cppreference.com/w/cpp/algorithm/set_difference. this should be "new_set.begin(), new_set.end(), old_set.begin(), old_set.end()"

File: src/mds/MDSRank.cc [hunk 4]
```diff
@@ -2656,56 +2666,68 @@ bool MDSRankDispatcher::kill_session(int64_t session_id, bool wait, std::strings
...
-      return false;
+      // Even though it wasn't us that removed it, kick our completion
+      // as the session has been removed.
+      if (on_killed) {
+        on_killed->complete(0);
+      }
     }
+  };
+
+  auto background_blacklist = [this, session_id, kill_mds_session, cmd](std::function<void ()> fn){
```
Comment (ukernel) [inline src/mds/MDSRank.cc]: kill_mds_session is not required here

**Comment by ukernel** [comment]
LGTM

**Comment by ukernel** [comment]
@jcsp  please fix Objecter::emit_blacklist_events

**Pull request merged**